Add support for "hardened" Adobe DRM & crypto code clean-up

a980e066a01 opened the pull request:
This PR adds support for the newer (""hardened"") Adobe DRM added with RMSDK 10.
It also cleans up the crypto code by removing support for anything but PyCryptodome and hashlib (faster than PyCryptodome). This actually fixes a few bugs on its own. You might want to review the two commits separately.

Closes #20, #25, #33, #45.

* The second commits **adds support for ""hardened"" Adobe DRM** and updates the documentation accordingly. In a nutshell, this so-called "hardened DRM" is just another layer of encryption on the encrypted key, using info that's fully available in `rights.xml`
* The first commit cleans up the crypto code and only retains support for PyCryptodome + hashlib, removing technical debt. I needed to do that to simplify "hardened" Adobe DRM support, but I figured that I could do that for the entire project: on Windows, openssl libs are most of the time not installed, and I think it's the same thing for Macs

I have tested my code on Amazon, ADE and Kobo DRMed books, but hopefully I haven't broken anything.

j-howell commented:
Thank you. It looks like a lot of work went into that.

PaulGJL commented:
This doesn't work for ADE books. Tried several times, with two different Calibre versions (4.13 and 5.12). Which version did you test the ADE books on? I have disabled old plugins, installed your newer version of the tool successfully and it can't still remove any Adobe  DRMs! please help.

ElleKayEm commented:
@PaulGJL You're using this patch on the newer Adobe DRM?

choc96208 commented:
@PaulGJL have you tried cloning this particular branch?

`git clone https://github.com/a980e066a01/DeDRM_tools.git --branch patch-hardened-drm`

then create a release with

`./make_release.py`

Source: [How do I clone a single branch in Git?](https://stackoverflow.com/questions/1778088/how-do-i-clone-a-single-branch-in-git)

noDRM commented:
Wow, thanks a lot for the new Adobe DRM support. I wonder if this is going to cause yet another DMCA claim to this repository...

I am going to take a look at the code and see if it works.

noDRM commented on DeDRM_plugin/ignoblekeyGenPassHash.py:
```diff
@@ -214,9 +139,8 @@ def generate_key(name, ccn):
-    both_sha = hashlib.sha1(name + ccn).digest()
-    aes = AES(ccn_sha, name_sha)
-    crypt = aes.encrypt(both_sha + (b'\x0c' * 0x0c))
+    both_sha = hashlib.sha1()(name + ccn).digest()
```
This line seems to be invalid and causes an exception, the first set of brackets should probably be removed.

a980e066a01 replied:
Oops - fixed

noDRM commented on DeDRM_plugin/ineptepub.py:
```diff
@@ -446,49 +231,43 @@ def isPassHashBook(inpath):
         except:
             return None
 
-def verify_book_key(bookkey):
```
Is there a particular reason you removed this function? Given that this plugin tries to use all available decryption keys, it's apparently possible that using a wrong key to decrypt the book key leads to garbage data, which this function is supposed to detect. Or is that redundant with your rewrite of how the decryption works (dropping OpenSSL support)?

See https://github.com/apprenticeharper/DeDRM_tools/issues/1862 for the reasons as to why this was added.

a980e066a01 replied:
> Or is that redundant with your rewrite of how the decryption works (dropping OpenSSL support)?

That is correct. The code that was using OpenSSL did not remove the padding, but the code that was (and is now) using PyCryptodome did. See: https://github.com/Legrandin/pycryptodome/blob/master/lib/Crypto/Cipher/PKCS1_v1_5.py . PyCryptodome removes the padding, hence the returned bookkey is always the proper length and doesn't start with `00 02`.

Two scenarios happen if the userkey or bookkey is incorrect: int(ciphertext) is >= than the modulus, in which case `ValueError` is raised; or the "sentinel" value is returned (I chose `None`).

noDRM commented on DeDRM_plugin/ineptepub.py:
```diff
@@ -518,32 +294,21 @@ def decryptBook(userkey, inpath, outpath):
-                # Verify key:
-                if len(bookkey) > 16:
-                    # Padded as per RSAES-PKCS1-v1_5
-                    if verify_book_key(bookkey):
```
Same comment as above - I'm not sure if removing `verify_book_key` is a good idea.

noDRM commented on DeDRM_plugin/ineptpdf.py:
```diff
@@ -1893,32 +1583,13 @@ def initialize_standard(self, password, docid, param):
         self.ready = True
         return
 
-    def verify_book_key(self, bookkey):
```
Same as for EPUB - is it a good idea to remove this?

johnhutnut replied:
> Same as for EPUB - is it a good idea to remove this?

Adobe Content Server 5 (ACS 5) DRM ，does any tools can decypt it ？ you can email me ab1307711@hotmail.com ，i will send you acsm file with adobe content server5 drm

noDRM replied:
This PR probably can. But please don't use review comments for support requests.

a980e066a01 replied:
As above.

noDRM reviewed (commented):
Just tested your changes and I was able to remove the DRM from an ADE3.0+ book. Great!
Though, I did notice one typo in your code that breaks some stuff, and I'm wondering why you removed the `verify_book_key` function? That was added for a reason, and I'm not sure if your changes to the encryption code are sufficient to the point where that wouldn't be needed anymore. See https://github.com/apprenticeharper/DeDRM_tools/issues/1862 for details as to why this was added.

johnhutnut commented:
i install it ，but fail calibre, version 5.38.0
错误： 未处理的异常: <b>ModuleNotFoundError</b>:No module named '__version'

calibre 5.38 [64bit]  embedded-python: True is64bit: True
Windows-10-10.0.22000 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.22000')
Python 3.8.5
Windows: ('10', '10.0.22000', '', 'Multiprocessor Free')
Interface language: zh_CN
Successfully initialized third party plugins: DeDRM (10, 0, 2) && Obok DeDRM (10, 0, 0)
Traceback (most recent call last):
  File "calibre\gui2\preferences\plugins.py", line 326, in add_plugin
  File "calibre\customize\ui.py", line 476, in add_plugin
  File "calibre\customize\ui.py", line 64, in load_plugin
  File "calibre\customize\zipplugin.py", line 290, in load
  File "polyglot\builtins.py", line 123, in reload
  File "importlib\__init__.py", line 169, in reload
  File "<frozen importlib._bootstrap>", line 604, in _exec
  File "calibre\customize\zipplugin.py", line 191, in exec_module
  File "calibre_plugins.dedrm.__init__", line 97, in <module>
ModuleNotFoundError: No module named '__version'

ElleKayEm commented:
@johnhutnut 
Looks like the plugin (with this patch) is not getting loaded?

noDRM commented:
@johnhutnut You are trying to load the plugin from source code - that's not supported. If you want to run versions that aren't released yet, use the make_release.py script to generate the proper archive.

johnhutnut commented:
can you release a new versions ，i am doctor，dont know much about python

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: noDRM ***@***.***>
发送时间: Sunday, March 6, 2022 11:57:09 PM
收件人: noDRM/DeDRM_tools ***@***.***>
抄送: johnhutnut ***@***.***>; Mention ***@***.***>
主题: Re: [noDRM/DeDRM_tools] Add support for "hardened" Adobe DRM & crypto code clean-up (PR #48)


@johnhutnut<https://github.com/johnhutnut> You are trying to load the plugin from source code - that's not supported. If you want to run versions that aren't released yet, use the make_release.py script to generate the proper archive.

―
Reply to this email directly, view it on GitHub<https://github.com/noDRM/DeDRM_tools/pull/48#issuecomment-1059988236>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AYC7MY4EZST4VFEQXPI4SQLU6TIVLANCNFSM5PVCQRHQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

johnhutnut commented:
> @johnhutnut You are trying to load the plugin from source code - that's not supported. If you want to run versions that aren't released yet, use the make_release.py script to generate the proper archive.

can you make a update version for me

noDRM commented:
I'm not going to create a release with untested code. Once the review of this PR is finished, the PR has been merged and the rest of the code is cleaned up a bit, there will be a new release.

johnhutnut commented:
i don't know how to generate the achive ，can you package one archive for ，thank you so much
________________________________
发件人: noDRM ***@***.***>
发送时间: Monday, March 7, 2022 12:06:51 AM
收件人: noDRM/DeDRM_tools ***@***.***>
抄送: johnhutnut ***@***.***>; Mention ***@***.***>
主题: Re: [noDRM/DeDRM_tools] Add support for "hardened" Adobe DRM & crypto code clean-up (PR #48)


I'm not going to create a release with untested code. Once the review of this PR is finished, the PR has been merged and the rest of the code is cleaned up a bit, there will be a new release.

―
Reply to this email directly, view it on GitHub<https://github.com/noDRM/DeDRM_tools/pull/48#issuecomment-1059990003>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AYC7MY4LCPS4TEVGBVURKH3U6TJZXANCNFSM5PVCQRHQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

johnhutnut commented:
i am doctor ，not a programer

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: noDRM ***@***.***>
发送时间: Monday, March 7, 2022 12:06:51 AM
收件人: noDRM/DeDRM_tools ***@***.***>
抄送: johnhutnut ***@***.***>; Mention ***@***.***>
主题: Re: [noDRM/DeDRM_tools] Add support for "hardened" Adobe DRM & crypto code clean-up (PR #48)


I'm not going to create a release with untested code. Once the review of this PR is finished, the PR has been merged and the rest of the code is cleaned up a bit, there will be a new release.

―
Reply to this email directly, view it on GitHub<https://github.com/noDRM/DeDRM_tools/pull/48#issuecomment-1059990003>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AYC7MY4LCPS4TEVGBVURKH3U6TJZXANCNFSM5PVCQRHQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

ElleKayEm commented:
> i am doctor ，not a programer 获取 Outlook for iOS<https://aka.ms/o0ukef>
> […](#)
> ________________________________ 发件人: noDRM ***@***.***> 发送时间: Monday, March 7, 2022 12:06:51 AM 收件人: noDRM/DeDRM_tools ***@***.***> 抄送: johnhutnut ***@***.***>; Mention ***@***.***> 主题: Re: [noDRM/DeDRM_tools] Add support for "hardened" Adobe DRM & crypto code clean-up (PR #48) I'm not going to create a release with untested code. Once the review of this PR is finished, the PR has been merged and the rest of the code is cleaned up a bit, there will be a new release. ― Reply to this email directly, view it on GitHub<[#48 (comment)](https://github.com/noDRM/DeDRM_tools/pull/48#issuecomment-1059990003)>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AYC7MY4LCPS4TEVGBVURKH3U6TJZXANCNFSM5PVCQRHQ>. Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. You are receiving this because you were mentioned.Message ID: ***@***.***>

Then you're better off waiting for the next release.

DarthSaddest commented:
> i don't know how to generate the achive ，can you package one archive for ，thank you so much

Install the zip from <https://we.tl/t-bNCKVOTyNx> as a plugin. Do NOT extract anything. [Dislcaimer: Verify_key functions are absent and you might have issues.]

johnhutnut commented:
> Install the zip from https://we.tl/t-bNCKVOTyNx as a plugin. Do NOT extract anything. [Dislcaimer: Verify_key functions are absent and you might have issues.]

thanks a lot ，work fine

tvphanatic commented:
> > Install the zip from https://we.tl/t-bNCKVOTyNx as a plugin. Do NOT extract anything. [Dislcaimer: Verify_key functions are absent and you might have issues.]
> 
> thanks a lot ，work fine

I just tried this with two ProQuest ebooks downloaded today, and it worked great on both; thanks!

PaulGJL commented:
Hi, so this patch does work for most ADE books now, but it has failed with a particular book. Seems like there's a problem with the code? Please check, thanks.

> DeDRM v10.0.2: Trying to decrypt The New Zealand Bill of Rights Act.pdf
> [37.58] splash screen hidden
> [37.58] Started up in 37.58 seconds with 12 books
> stdout+stderr from file dialog helper: [b'', b'']
> piped data from file dialog helper: [b'\xc3\x9d\xa9\xa87\xb3\xfbe\xb6\xcb;[\xa9S\xde\xdd\xa4\x11^\xadC8\xe6\xf3\xd1Q\x87\xdb\xae\x08E\xc6', b'C:\\Users\\tgots\\Documents\\My Digital Editions\\The New Zealand Bill of Rights Act.pdf']
> DeDRM v10.0.2: The New Zealand Bill of Rights Act.pdf is a PDF ebook with encryption EBX_HANDLER
> DeDRM v10.0.2: The New Zealand Bill of Rights Act.pdf is a PDF ebook (EBX) for UUID 86029e8d-c7be-4607-949c-fd6098c3aadf
> DeDRM v10.0.2: Trying encryption key default_key
> error writing pdf: Invalid object number: objid=13803
> Traceback (most recent call last):
>   File "C:\Users\tgots\AppData\Roaming\calibre\plugins\DeDRM.zip\ineptpdf.py", line 1804, in getobj
> IndexError: list index out of range
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "C:\Users\tgots\AppData\Roaming\calibre\plugins\DeDRM.zip\ineptpdf.py", line 2274, in decryptBook
>   File "C:\Users\tgots\AppData\Roaming\calibre\plugins\DeDRM.zip\ineptpdf.py", line 2102, in dump
>   File "C:\Users\tgots\AppData\Roaming\calibre\plugins\DeDRM.zip\ineptpdf.py", line 1806, in getobj
> ineptpdf.PDFSyntaxError: Invalid object number: objid=13803
> DeDRM v10.0.2: Failed to decrypt with key default_key after 7.5 seconds
> DeDRM v10.0.2: Looking for new default Adobe Digital Editions Keys after 7.5 seconds
> Found 1 keys
> DeDRM v10.0.2: Trying a new default key
> error writing pdf: Invalid object number: objid=13803
> Traceback (most recent call last):
>   File "C:\Users\tgots\AppData\Roaming\calibre\plugins\DeDRM.zip\ineptpdf.py", line 1804, in getobj
> IndexError: list index out of range
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "C:\Users\tgots\AppData\Roaming\calibre\plugins\DeDRM.zip\ineptpdf.py", line 2274, in decryptBook
>   File "C:\Users\tgots\AppData\Roaming\calibre\plugins\DeDRM.zip\ineptpdf.py", line 2102, in dump
>   File "C:\Users\tgots\AppData\Roaming\calibre\plugins\DeDRM.zip\ineptpdf.py", line 1806, in getobj
> ineptpdf.PDFSyntaxError: Invalid object number: objid=13803
> DeDRM v10.0.2: Failed to decrypt with new default key after 11.7 seconds
> DeDRM v10.0.2: Finished after 11.7 seconds
> Traceback (most recent call last):
>   File "calibre\customize\ui.py", line 428, in get_file_type_metadata
>   File "calibre\customize\builtins.py", line 347, in get_metadata
>   File "calibre\ebooks\metadata\pdf.py", line 126, in get_metadata
> ValueError: Could not read info dict from PDF
> Syntax Error: Couldn't find the 'EBX_HANDLER' security handler
> 
> pdfinfo errored out with return code: 1

kokowang8910 commented:
Thanks a lot for the effort. it works! awesome!

> Install the zip from https://we.tl/t-bNCKVOTyNx as a plugin. Do NOT extract anything. [Dislcaimer: Verify_key functions are absent and you might have issues.]

usure commented:
> > i don't know how to generate the achive ，can you package one archive for ，thank you so much
> 
> Install the zip from https://we.tl/t-bNCKVOTyNx as a plugin. Do NOT extract anything. [Dislcaimer: Verify_key functions are absent and you might have issues.]

Doesn't seem to work for me. I'm on Windows 7 and using a Calibre version prior to 5.38.0 -- might that be the problem?

ElleKayEm commented:
@usure 
If it's calibre before 5.x, then that's probably it.

usure commented:
> @usure If it's calibre before 5.x, then that's probably it.

Thanks. Makes sense. Will try on my Win 8 laptop.

noDRM commented:
The goal is for this plugin to work with Calibre 4.x, too, at least for the most common DRM (Adobe). If it doesn't, can you post a debug log?

boan-anbo commented:
@a980e066a01 Thanks so much. It works for me as well

ableeker commented:
I've implemented these changes, and I've tried to convert some ebooks. Most worked, but the B&N ebooks I have didn't.

I had a look at the new code, and I noticed that the book key that had been decrypted by this new code was 32 bytes long, while the bookkey from the current code was 16 bytes long. I noticed that the current code checks if the book key is longer than 16 bytes, and if that's the case, it only uses the last 16 bytes. This code has been omitted, but still seems needed, at least to convert the ebooks I have. Actually, this check is only needed if the book key can be shorter than 16 bytes, but maybe that doesn't happen anyway. In that case it may suffice to cut the extra bytes.

If the book key always is 16 bytes, or longer, the correct book key can be generated by changing the line:

`bookkey = unpad(AES.new(key, AES.MODE_CBC, b'\x00'*16).decrypt(bookkey), 16)`

to:

`bookkey = unpad(AES.new(key, AES.MODE_CBC, b'\x00'*16).decrypt(bookkey), 16)[-16:]`

If the book key can be shorter than 16 bytes, this line doesn't have to be modified, and the following lines from the current code should be placed after the line:

```
if len(bookkey) > 16:
    bookkey = bookkey[-16:]
```

When the new code is modified this way, my B&N ebooks convert successfully.

a980e066a01 commented:
Thanks for letting me know...

Can you remove the `unpad` and print `hexlify(bookkey)` for these books please?

ableeker commented:
One ebook has book key:

`b'\xa0"?\xb3YN\xea\x12\x9c\xb1\xbe\x00c\xa8V\xed\xbe,Uz4\xb7\xa9\xe6p\xeb\x83\x8c\xe8\r\xd2H\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10'`
`b'a0223fb3594eea129cb1be0063a856edbe2c557a34b7a9e670eb838ce80dd24810101010101010101010101010101010'`

The other ebook has book key:

`b'\x06\x19\xb2<pD\xdb\xba\xbfgM\xf6"@\xb3_|\x1f\xc7x]I<\xc5\x1d\x84\xee\x9e>\xd3Y0\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10\x10'`
`b'0619b23c7044dbbabf674df62240b35f7c1fc7785d493cc51d84ee9e3ed3593010101010101010101010101010101010'`

The book key needed to convert the first ebook is:

`b'\xbe,Uz4\xb7\xa9\xe6p\xeb\x83\x8c\xe8\r\xd2H'`
`b'be2c557a34b7a9e670eb838ce80dd248'`

The book key needed to convert the second ebook is:
`b'|\x1f\xc7x]I<\xc5\x1d\x84\xee\x9e>\xd3Y0'`
`b'7c1fc7785d493cc51d84ee9e3ed35930'`

a980e066a01 commented:
Thanks. The commits I've just force-pushed should have fixed that. Also added a few things I've missed.

LittleMiller commented:
I tried to open the link (https://we.tl/t-bNCKVOTyNx) to download the file, but it said the transfer had expired. Can one of you send it again?

aadam1x commented:
> > i don't know how to generate the achive ，can you package one archive for ，thank you so much
> 
> Install the zip from https://we.tl/t-bNCKVOTyNx as a plugin. Do NOT extract anything. [Dislcaimer: Verify_key functions are absent and you might have issues.]

Hi this is expired now, do you mind sharing again? :))

stopregionblocking commented:
> > i don't know how to generate the achive ，can you package one archive for ，thank you so much
> 
> Install the zip from https://we.tl/t-bNCKVOTyNx as a plugin. Do NOT extract anything. [Dislcaimer: Verify_key functions are absent and you might have issues.]

The download has expired. Reuploaded to https://bayfiles.com/vbs9s0O2x4/DeDRM_plugin_zip as it's already being requested.

LittleMiller commented:
Thank you @stopregionblocking. Works like a charm

aadam1x commented:
@stopregionblocking legend! Works perfectly thanks.

maksimumeffort commented:
Hey guys I'm getting an error using [@stopregionblocking](https://github.com/stopregionblocking) 's zip file in calibre. I'm getting the following `ModuleNotFoundError:No module named '__version'`
Checked inside the zip file and can confirm it has a file named '__version' 
compared the zip file with second commit and they are identical. not sure what's the problem
Currently running Calibre 5.0.1. Would updating calibre solve it? 

Any help greatly appreciated.
Cheers
-A

a980e066a01 commented:
@noDRM Np, go ahead

maksimumeffort commented:
UPDATE: tried updating to calibre 5.1. Same issue.
Anyone got a solution?

ElleKayEm commented:
@maksimumeffort 
Are you unable to load the plugin at all?  I'm able to load the plugin from stopregionblocking's link.  But I don't have any "hardened" DRM ADE files to test with. (No problem with the old Adobe DRM though.)  Sounds like you're having the same issue as johnhutnut above who was trying to load from the source code instead of a proper archive.  Shouldn't have long to wait for the official release though.

ableeker commented:
Just zipping the source won't create an archive. I think you need to have Python installed on your computer first. Then you can run make 'python make_release.py 10.0.3' to create an archive.

stopregionblocking commented:
> @maksimumeffort I'm able to load the plugin from stopregionblocking's link. But I don't have any "hardened" DRM ADE files to test with.

ProQuest EBook Central recently switched to the "hardened" DRM - try it on something from there.

ElleKayEm commented:
> ProQuest EBook Central recently switched to the "hardened" DRM - try it on something from there.

Don't have access to that.

maksimumeffort commented:
@stopregionblocking you're right on the money, I'm trying to deDRM a book I got through ProQuest right now and it's a total pain. 

@ableeker I'm not sure what you mean. I was under the impression that it's a straight forward install (like the one shown @ https://epubor.com/free-kindle-drm-removal-calibre-plugin.html) I do have python installed though, so I can run that command on the file I downloaded from @stopregionblocking 's link.

Also do I need to have a previous deDRM stable plugin version installed before I install the "hardened" plugin or is the "hardened" plugin just a copy with added decryption scripts?

ElleKayEm commented:
@maksimumeffort 
If you're getting that error, then you seem to be trying to load the zipped up DeDRM_plugin folder from the source code...not the archive at stopregionblocking's link.  Redownload, naming it something different, and try that.

maksimumeffort commented:
@ElleKayEm Tried it your way just now, still same error
`ModuleNotFoundError:No module named -__version'`

ElleKayEm commented:
@maksimumeffort Are you using the official calibre from here: https://calibre-ebook.com/download

maksimumeffort commented:
@ElleKayEm Yep. Running version 5.1 right now

ElleKayEm commented:
@maksimumeffort Just checking because there can be issues with old, buggy versions that come with Linux distros and aren't using calibre's embedded Python.

I'm using the current calibre 5.39 and I did already have the last DeDRM 10.0.2 release loaded.

noDRM commented:
Not sure where that `__version` error might be coming from, but since this PR doesn't change anything related to that at all, it must be caused by one of my prior changes between 10.0.2 and the current master, when `__version` has been introduced. I'm going to merge this PR as-is, then try to fix PDFs with V3 encryption, and then make a new full beta release for people to test.

ableeker commented:
I get the error `ModuleNotFound:No module named '__version'` if I zip the current source files, and try to add this zip as a plug-in in Calibre. That used to work, but it doesn't anymore. To create a plug-in from the source files, I now use the command:

`python makerelease.py 10.0.3`.

The Python program `make_release.py` is provided in the root of the folder that contains the source. When you run this program, it creates a file DeDRM_tools_10.0.3.zip. This file contains a zip named DeDRM_plugin.zip. This zip file (don't unzip it!) can be loaded in Calibre as the desired plug-in.

maksimumeffort commented:
@noDRM Hey mate, I just ran the bet in my calibre (Calibre 5.1) and now I get the following error:

`InvalidPlugin:The plugin in 'users/'username'/Desktop/DeDRM_tools_beta_2022-03-18.zip is invalid. It does not contain a top-level __init__.py file `

I'll try to update to latest calibre and see if that solves the problem. If not I'll try the solution above.

@ableeker did you run that command while in directory of the cloned repo?

maksimumeffort commented:
UPDATE: just ran the beta with latest calibre (Calibre 5.39) and I'm getting the same error message, renaming the beta did not help.

ElleKayEm commented:
@maksimumeffort Unzip DeDRM_tools_beta_2022-03-18.zip and load just DeDRM_plugin.zip.

noDRM commented:
> I get the error `ModuleNotFound:No module named '__version'` if I zip the current source files, and try to add this zip as a plug-in in Calibre. That used to work, but it doesn't anymore. To create a plug-in from the source files, ...

Correct, just zipping up the source folder and adding that to Calibre no longer works (due to some of the code changes I made after v10.0.2). I'll try to make it print a more useful error message in that case. 
But @maksimumeffort seems to have gotten that error message even with the prepared archive that someone posted a link to. I didn't look too closely at that archive but it seems to have been generated with the release script. That's why II'm having him try my beta ZIP (which, as ElleKayEm already said needs to be extracted first ...) and see if that makes the `__version` error go away.

ElleKayEm commented:
For whatever it's worth, I didn't get that error with the linked archive and no one else here reported that issue either.  The only other person with that error was zipping up the source folder.  They were fine once a properly generated archive was posted.

noDRM commented:
@a980e066a01 Was there a particular reason you replaced the `unpad` function with a call to PyCrypto? That's working fine on Python 3, but on Python 2 (Calibre < 5) this fails because unpad doesn't seem to exist there. I've reverted these changes and added back the old unpad code to restore compatibility with the old Calibre.

johnhutnut commented:
i have clean up all old calibre and python，install newest version of them ，work fine in win11

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: noDRM ***@***.***>
发送时间: Saturday, March 19, 2022 5:16:46 PM
收件人: noDRM/DeDRM_tools ***@***.***>
抄送: johnhutnut ***@***.***>; Mention ***@***.***>
主题: Re: [noDRM/DeDRM_tools] Add support for "hardened" Adobe DRM & crypto code clean-up (PR #48)


@a980e066a01<https://github.com/a980e066a01> Was there a particular reason you replaced the unpad function with a call to PyCrypto? That's working fine on Python 3, but on Python 2 (Calibre < 5) this fails because unpad doesn't seem to exist there. I've reverted these changes and added back the old unpad code to restore compatibility with the old Calibre.

―
Reply to this email directly, view it on GitHub<https://github.com/noDRM/DeDRM_tools/pull/48#issuecomment-1072974862>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AYC7MYZB5EF365O2DU2KDC3VAWLP5ANCNFSM5PVCQRHQ>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

ElleKayEm commented:
calibre has its own embedded python.

dskfsdfksdfk commented:
@noDRM 
> @dskfsdfksdfk Can you try this beta version, please. This should hopefully fix the bug with the V3-encrypted PDF files (but I can't test that myself as I don't have such a file), and it will add some more logging for that __version issue that hopefully helps me track this down.
> 
> 

Works flawlessly! Thank you so much.

a980e066a01 commented:
> @a980e066a01 Was there a particular reason you replaced the `unpad` function with a call to PyCrypto? That's working fine on Python 3, but on Python 2 (Calibre < 5) this fails because unpad doesn't seem to exist there. I've reverted these changes and added back the old unpad code to restore compatibility with the old Calibre.

Refactoring, mostly, and I'm not sure all the reimpls handled all corner cases correctly.

Looks like unpad is not present in PyCrypto (but is in PyCryptodome). Urgh.

ZolaLa9 commented:
Thank you @noDRM and @a980e066a01 for all your work on this. I don't have any books with ADE hardened DRM to test, but I've been doing quick tests on other formats to make sure everything else still works. 

**@noDRM — the latest changes seems to have broken decryption on Kindle for Mac.** 

I downloaded [DeDRM_tools_beta_2022-03-18 that you posted above](https://github.com/noDRM/DeDRM_tools/pull/48#issuecomment-1072589825) and **that plugin works fine for me, no problems encountered.** 

But I did **make_release.py on latest Master (with last commit being 2d51005) and resulting plugin does _not_ work on Kindle for Mac (but it still works for Kindle books downloaded for transfer via USB and older Adobe DRM).**

I am on macOS 10.15.7 using Calibre 5.39.0 and Kindle for Mac 1.31.1 with KFX disabled via chmod -x. As I said [DeDRM_tools_beta_2022-03-18](https://github.com/noDRM/DeDRM_tools/pull/48#issuecomment-1072589825) works fine, but plugin using latest Master (up to commit 2d51005) no longer works. 

calibre Debug log
calibre 5.39  embedded-python: True is64bit: True
macOS-10.15.7-x86_64-i386-64bit Darwin ('64bit', '')
('Darwin', '19.6.0', 'Darwin Kernel Version 19.6.0: Tue Feb 15 21:39:11 PST 2022; root:xnu-6153.141.59~1/RELEASE_X86_64')
Python 3.8.5
OSX: ('10.15.7', ('', '', ''), 'x86_64')
Interface language: None
Successfully initialized third party plugins: DeACSM (0, 0, 15) && Gather KFX-ZIP (from KFX Input) (1, 47, 0) && DeDRM (10, 0, 2, 1) && Package KFX (from KFX Input) (1, 47, 0) && Diaps Editing Toolbag (0, 3, 7) && EpubCheck (0, 2, 4) && KFX metadata reader (from KFX Input) (1, 47, 0) && KFX Input (1, 47, 0) && Set KFX metadata (from KFX Output) (1, 60, 0) && KFX Output (1, 60, 0) && KindleUnpack - The Plugin (0, 83, 1) && Modify ePub (1, 7, 0) && Quality Check (1, 12, 0) && ScrambleEbook (0, 4, 5)
calibre 5.39  embedded-python: True is64bit: True
macOS-10.15.7-x86_64-i386-64bit Darwin ('64bit', '')
('Darwin', '19.6.0', 'Darwin Kernel Version 19.6.0: Tue Feb 15 21:39:11 PST 2022; root:xnu-6153.141.59~1/RELEASE_X86_64')
Python 3.8.5
OSX: ('10.15.7', ('', '', ''), 'x86_64')
Interface language: None
Successfully initialized third party plugins: DeACSM (0, 0, 15) && Gather KFX-ZIP (from KFX Input) (1, 47, 0) && DeDRM (10, 0, 2, 1) && Package KFX (from KFX Input) (1, 47, 0) && Diaps Editing Toolbag (0, 3, 7) && EpubCheck (0, 2, 4) && KFX metadata reader (from KFX Input) (1, 47, 0) && KFX Input (1, 47, 0) && Set KFX metadata (from KFX Output) (1, 60, 0) && KFX Output (1, 60, 0) && KindleUnpack - The Plugin (0, 83, 1) && Modify ePub (1, 7, 0) && Quality Check (1, 12, 0) && ScrambleEbook (0, 4, 5)
devicePixelRatio: 2.0
logicalDpi: 72.0 x 72.0
physicalDpi: 91.00000136668288 x 91.0000013666829
Using calibre Qt style: True
[0.00] Starting up...
[0.03] Showing splash screen...
[0.26] splash screen shown
[0.26] Initializing db...
[0.27] db initialized
[0.27] Constructing main UI...
[1.29] main UI initialized...
[1.29] Hiding splash screen
Starting QuickView
Gather KFX-ZIP (from KFX Input) 1.47.0: Importing /Users/Z/Downloads/B0XXXXXXXX_EBOK.azw
Gather KFX-ZIP (from KFX Input): File is not KFX format
DeDRM v10.0.2.1: Trying to decrypt B0XXXXXXXX_EBOK.azw
Using Library AlfCrypto Python
MobiDeDrm v1.1.
Copyright © 2008-2020 The Dark Reverser, Apprentice Harper et al.
Decrypting Kindle Format 8 ebook: This is the Book Title
Found 2 keys to try after 0.1 seconds
Crypto Type is: 2
DeDRM v10.0.2.1: Failed to decrypt with error: No key found in 2 PIDs tried.
DeDRM v10.0.2.1: Looking for new default Kindle Key after 0.1 seconds
Found k4Mac kinf2018 file: /Users/Z/Library/Application Support/Kindle/storage/.kinf2018
trying username  b'Z'  on file  /Users/Z/Library/Application Support/Kindle/storage/.kinf2018
trying IDString: b'xxx'
trying IDString: b'xxx'
[4.23] splash screen hidden
[4.23] Started up in 4.23 seconds with 0 books
Decrypted key file using IDString 'xxx' and UserName 'Z'
DeDRM v10.0.2.1: Found 1 new key
DeDRM v10.0.2.1: Ultimately failed to decrypt after 0.6 seconds. Read the FAQs at noDRM's repository: https://github.com/noDRM/DeDRM_tools/blob/master/FAQs.md
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "calibre_plugins.dedrm.__init__", line 957, in KindleMobiDecrypt
    book = k4mobidedrm.GetDecryptedBook(path_to_ebook,kindleDatabases,androidFiles,serials,pids,self.starttime)
  File "/Users/Z/Library/Preferences/calibre/plugins/DeDRM.zip/k4mobidedrm.py", line 261, in GetDecryptedBook
    mb.processBook(totalpids)
  File "/Users/Z/Library/Preferences/calibre/plugins/DeDRM.zip/mobidedrm.py", line 501, in processBook
    raise DrmException("No key found in {0:d} PIDs tried.".format(len(goodpids)))
mobidedrm.DrmException: No key found in 2 PIDs tried.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "calibre/customize/ui.py", line 175, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 1058, in run
    decrypted_ebook = self.KindleMobiDecrypt(path_to_ebook)
  File "calibre_plugins.dedrm.__init__", line 1009, in KindleMobiDecrypt
    raise DeDRMError("{0} v{1}: Ultimately failed to decrypt after {2:.1f} seconds. Read the FAQs at noDRM's repository: https://github.com/noDRM/DeDRM_tools/blob/master/FAQs.md".format(PLUGIN_NAME, PLUGIN_VERSION,time.time()-self.starttime))
calibre_plugins.dedrm.DeDRMError: DeDRM v10.0.2.1: Ultimately failed to decrypt after 0.6 seconds. Read the FAQs at noDRM's repository: https://github.com/noDRM/DeDRM_tools/blob/master/FAQs.md
Added This is the Book Title to db in: 0.0
Added 1 books in 1.4 seconds

Let me know if there's anything more I can do to help. Thank you

noDRM commented:
Interesting, thanks for testing. Looks like one of today's commits broke that, then. I'll take a look at these again and try a Kindle eBook myself to test.

ZolaLa9 commented:
Ack! Sorry. I just opened a new issue for this thinking to delete this comment since it's not really related to this PR. I will delete the new issue.

And thank you for looking into this!

agentxan commented:
Trying the plugin after each commit the kindle book start giving an error on "(Hopefully) fix WineGetKeys for Kindle Package plugin #52" if you do not already have the keys imported. 

If I try "Add back unpad to fix Python2 support Package plugin #51" the keys are imported and the drm is removed . If I then install "(Hopefully) fix WineGetKeys for Kindle Package plugin #52"  it removes the drm but if I remove the imported keys it does not remove the drm

maksimumeffort commented:
@ElleKayEm It worked! Thanks heaps
@noDRM It's working thanks so much for your help and the zip file

All I had to do was download the latest calibre (calibre 5.39) download @noDRM 's zip file, unzip it and select DeDRM_plugin.zip file inside when installing plugin. 
I'm currently running macOS version 11.5.1

Hope this info helps anyone who's struggling.

Cheers, 
-A

ZolaLa9 commented:
> Trying the plugin after each commit the kindle ebooks start giving an error on "(Hopefully) fix WineGetKeys for Kindle Package plugin 52" if you do not already have the keys imported.
> 
> If I try "Add back unpad to fix Python2 support Package plugin 51" the keys are imported and the drm is removed . If I then install "(Hopefully) fix WineGetKeys for Kindle Package plugin 52" it removes the drm but if I remove the imported keys it does not remove the drm

Awesome. Problem is definitely in "(Hopefully) fix WineGetKeys for Kindle" bb17068.

I downloaded latest Master (up to commit 2d51005), replaced ```__init__.py``` (which is from "(Hopefully) fix WineGetKeys for Kindle" bb17068) with ```__init__.py``` from "Add back unpad to fix Python2 support" b283777, ran make_release.py, and clean install of the resulting plugin works on Kindle for Mac. (I don't know enough about code to go further than that.)

noDRM commented:
Tricky issue because the log you posted above doesn't contain useful error messages ... can you try 726d72217e0d6bb8229732a35fa85275a3b629f3 (latest master) and see if that fixes it?

agentxan commented:
Using the latest master I still get the error.

`calibre Debug log
calibre 5.39 [64bit] Portable embedded-python: True is64bit: True
Windows-10-10.0.22000 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.22000')
Python 3.8.5
Windows: ('10', '10.0.22000', '', 'Multiprocessor Free')
Interface language: None
Successfully initialized third party plugins: DeDRM (10, 0, 2)
calibre 5.39 [64bit] Portable embedded-python: True is64bit: True
Windows-10-10.0.22000 Windows ('64bit', 'WindowsPE')
('Windows', '10', '10.0.22000')
Python 3.8.5
Windows: ('10', '10.0.22000', '', 'Multiprocessor Free')
Interface language: None
Successfully initialized third party plugins: DeDRM (10, 0, 2)
Turning on automatic hidpi scaling
devicePixelRatio: 1.0
logicalDpi: 120.0 x 120.0
physicalDpi: 141.7674418604651 x 142.13471502590673
Using calibre Qt style: True
[0.00] Starting up...
[0.00] Showing splash screen...
[0.52] splash screen shown
[0.52] Initializing db...
[0.52] db initialized
[0.52] Constructing main UI...
[1.02] main UI initialized...
[1.02] Hiding splash screen
Starting QuickView
DeDRM v10.0.2: Trying to decrypt B079VDMTD5_EBOK.azw
Using Library AlfCrypto DLL/DYLIB/SO
MobiDeDrm v1.1.
Copyright © 2008-2020 The Dark Reverser, Apprentice Harper et al.
Decrypting Kindle Format 8 ebook: Regen (Penchant for Trouble Book 1)
Found 0 keys to try after 0.1 seconds
Crypto Type is: 2
DeDRM v10.0.2: Failed to decrypt with error: No key found in 0 PIDs tried.
DeDRM v10.0.2: Looking for new default Kindle Key after 0.1 seconds
searching for kinfoFiles in C:\Users\jbean\AppData\Local
Found K4PC 1.25+ kinf2018 file: C:\Users\jbean\AppData\Local\Amazon\Kindle\storage\.kinf2018
Found K4PC 1.9+ kinf2011 file: C:\Users\jbean\AppData\Local\Amazon\Kindle\storage\.kinf2011
Decrypted key file using IDString '2825280726' and UserName 'jfb'
Decrypted key file using IDString '2825280726' and UserName 'jfb'
DeDRM v10.0.2: Found 2 new keys
MobiDeDrm v1.1.
Copyright © 2008-2020 The Dark Reverser, Apprentice Harper et al.
Decrypting Kindle Format 8 ebook: Regen (Penchant for Trouble Book 1)
DeDRM v10.0.2: Ultimately failed to decrypt after 0.3 seconds. Read the FAQs at noDRM's repository: https://github.com/noDRM/DeDRM_tools/blob/master/FAQs.md
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "calibre_plugins.dedrm.__init__", line 957, in KindleMobiDecrypt
  File "C:\Calibre5Portable\Calibre Portable\Calibre Settings\plugins\DeDRM.zip\k4mobidedrm.py", line 261, in GetDecryptedBook
  File "C:\Calibre5Portable\Calibre Portable\Calibre Settings\plugins\DeDRM.zip\mobidedrm.py", line 501, in processBook
mobidedrm.DrmException: No key found in 0 PIDs tried.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "calibre\customize\ui.py", line 175, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 1058, in run
  File "calibre_plugins.dedrm.__init__", line 1009, in KindleMobiDecrypt
calibre_plugins.dedrm.DeDRMError: DeDRM v10.0.2: Ultimately failed to decrypt after 0.3 seconds. Read the FAQs at noDRM's repository: https://github.com/noDRM/DeDRM_tools/blob/master/FAQs.md
[2.45] splash screen hidden
[2.45] Started up in 2.45 seconds with 0 books
Added Regen (Penchant for Trouble Book 1) to db in: 0.1
Added 1 booWorker Launch took: 0.00 seconds
`

ElleKayEm commented:
@agentxan Were you also unable to decrypt Kindle for PC books with the current release?  I see DeDRM finds a key for both an older and newer version of Kindle for PC.  Was this book downloaded with the current installation of Kindle for PC on this on this machine?

noDRM commented:
He wrote yesterday that if he reverts bb170688ba67557de2400f377ab907b991437b30 then the issue is gone, so I'm assuming his local setup is correct. 

I'm going to try and set up an old K4PC on my machine in a couple days to test this myself, that's probably easier given that the logs don't indicate any errors.

agentxan commented:
The book was downloaded with Kindle for PC 1.17.1.  If I go back to [Add back unpad to fix Python2 support](https://github.com/noDRM/DeDRM_tools/commit/b283777c0a429e5f0ec322e5c27d0579ca96dcbb) it removes the drm without errors and imports the keys. If I then try a newer version after that commit it removes the drm without error but if I remove the keys from the plugin settings it then starts to show the error.

noDRM commented:
Okay, now I tested it with my own Windows VM, 76ce6d9c5c813b57d37173280cf3800b90530970 finally fixes this for me, would you mind testing again?

agentxan commented:
That fixed the problem for me. Thank you

ZolaLa9 commented:
@noDRM — thank you! Latest master (76ce6d9) works for me on macOS

JSWolf commented:
> > Install the zip from https://we.tl/t-bNCKVOTyNx as a plugin. Do NOT extract anything. [Dislcaimer: Verify_key functions are absent and you might have issues.]
> 
> thanks a lot ，work fine

Can tyou please post again? The upload has since expired. Thanks.

ElleKayEm commented:
@JSWolf If you don't want to run makerelease.py yourself, the one noDRM posted in this comment is more recent:  https://github.com/noDRM/DeDRM_tools/pull/48#issuecomment-1072589825

ElleKayEm commented:
@x0thrmc-exe You've tried with the current code (including what was merged from this pull request)?  Not just the last 10.0.2 release?

x0thrmc-exe commented:
> @x0thrmc-exe You've tried with the current code (including what was merged from this pull request)? Not just the last 10.0.2 release?

I am using the version from 2022-03-18 that was [linked above](https://github.com/noDRM/DeDRM_tools/pull/48#issuecomment-1072589825) in this thread. Is there a later one?

ElleKayEm commented:
No, not unless you run makerelease.py yourself.

ElleKayEm commented:
Hopefully @noDRM or someone can take a look at your last log and figure something out.

x0thrmc-exe commented:
Looking at other peoples error logs, I seem to be the only one that has anything like 

> ValueError: string argument should contain only ASCII characters

And I wonder if this could be due to the metadata of the book containing non-ASCII characters. It's a bilingual book with both Chinese and English. Although, all other similar books from the same source, author, and publisher have no problem.

x0thrmc-exe commented:
> Looking at the error code, this is unlikely to be caused by the book name. Do you get this error for multiple PDF books or just one? Or did you just test one?

Only one book had this problem. All other books from the same source, including other books in the series had no problems at all. 

> 
> This looks like the account key in the plugin settings is corrupted. Did you manually import that key, or did the plugin auto-discover it? Can you try removing the ADE key from the plugin settings (make a backup first, just in case) and try again, see if it auto-discovers the correct one?

I have tried both manually importing and auto-discovering, and on different platforms and computers. I have already tried removing and auto-discovering, but I will try once more.

ElleKayEm commented:
So this book is tied to the same Adobe ID as the ones which successfully decrypt?

x0thrmc-exe commented:
yes

> So this book is tied to the same Adobe ID as the ones which successfully decrypt?

ALonX1 commented:
> @maksimumeffort @dskfsdfksdfk Can you try this beta version, please. This should hopefully fix the bug with the V3-encrypted PDF files (but I can't test that myself as I don't have such a file), and it will add some more logging for that __version issue that hopefully helps me track this down.
> 
> [DeDRM_tools_beta_2022-03-18.zip](https://github.com/noDRM/DeDRM_tools/files/8306286/DeDRM_tools_beta_2022-03-18.zip)

Working fine on a ProQuest PDF tested today

stillhope commented:
Hi, in [DeDRM_tools_beta_2022-03-18.zip] it appears the code can be run without calibre, standalone
in file standalone/__init__.py 
says
# CLI interface for the DeDRM plugin (useable without Calibre, too)

To run as a standalone, what command to I type in terminal?  and do i need to be in a specific directory?
Thankyou for any help

noDRM commented:
This has nothing to do with this PR...
Same as #87, running the plugin standalone isn't possible yet. I started implementing it which is what you see in that file, but support for that isn't complete yet.

TheWranger commented:
> Looking at other peoples error logs, I seem to be the only one that has anything like
> 
> > ValueError: string argument should contain only ASCII characters
> 
> And I wonder if this could be due to the metadata of the book containing non-ASCII characters. It's a bilingual book with both Chinese and English. Although, all other similar books from the same source, author, and publisher have no problem.

Hey, I'm experiencing this error as well. I have a PDF from ProQuest I'm attempting to remove the DRM of, that won't download in ADE 2.0 so I have to use 4.5, and after attempting to use the new version of DeDRM linked at https://github.com/noDRM/DeDRM_tools/pull/48#issuecomment-1072589825

https://www.toptal.com/developers/hastebin/efufevujit.sql

ht55ght55 commented:
Ohh thank you, thank you!!!  I almost gave up again.

TheWranger commented:
> > Looking at other peoples error logs, I seem to be the only one that has anything like
> > > ValueError: string argument should contain only ASCII characters
> > 
> > 
> > And I wonder if this could be due to the metadata of the book containing non-ASCII characters. It's a bilingual book with both Chinese and English. Although, all other similar books from the same source, author, and publisher have no problem.
> 
> Hey, I'm experiencing this error as well. I have a PDF from ProQuest I'm attempting to remove the DRM of, that won't download in ADE 2.0 so I have to use 4.5, and after attempting to use the new version of DeDRM linked at [#48 (comment)](https://github.com/noDRM/DeDRM_tools/pull/48#issuecomment-1072589825)
> 
> https://www.toptal.com/developers/hastebin/efufevujit.sql

Here is another book which is still blocked by DRM with the same error. This one was downloaded from the Internet Archive.
https://www.toptal.com/developers/hastebin/popubojoyi.sql

TheWranger commented:
> > > Looking at other peoples error logs, I seem to be the only one that has anything like
> > > > ValueError: string argument should contain only ASCII characters
> > > 
> > > 
> > > And I wonder if this could be due to the metadata of the book containing non-ASCII characters. It's a bilingual book with both Chinese and English. Although, all other similar books from the same source, author, and publisher have no problem.
> > 
> > 
> > Hey, I'm experiencing this error as well. I have a PDF from ProQuest I'm attempting to remove the DRM of, that won't download in ADE 2.0 so I have to use 4.5, and after attempting to use the new version of DeDRM linked at [#48 (comment)](https://github.com/noDRM/DeDRM_tools/pull/48#issuecomment-1072589825)
> > https://www.toptal.com/developers/hastebin/efufevujit.sql
> 
> Here is another book which is still blocked by DRM with the same error. This one was downloaded from the Internet Archive. https://www.toptal.com/developers/hastebin/popubojoyi.sql



> > Looking at other peoples error logs, I seem to be the only one that has anything like
> > > ValueError: string argument should contain only ASCII characters
> > 
> > 
> > And I wonder if this could be due to the metadata of the book containing non-ASCII characters. It's a bilingual book with both Chinese and English. Although, all other similar books from the same source, author, and publisher have no problem.
> 
> Hey, I'm experiencing this error as well. I have a PDF from ProQuest I'm attempting to remove the DRM of, that won't download in ADE 2.0 so I have to use 4.5, and after attempting to use the new version of DeDRM linked at [#48 (comment)](https://github.com/noDRM/DeDRM_tools/pull/48#issuecomment-1072589825)
> 
> https://www.toptal.com/developers/hastebin/efufevujit.sql



> I have been struggling with DRM on a single ebook for over a week now. Calibre: 5.41 in Win10 VM with ADE 4.5 installed. and Calibre 5.25 in Linux, also tried on OSX DeDRM: 10.0.2 beta from 2022-03-18
> 
> I have been successful with every other PDF and EPUB from multiple sources. Also tried many from the same source as the problem ebook. I only have an error on this PDF so I can not figure out what the issue is.
> 
> > Failed to convert book: C:\Users*redacted my name*\Calibre Library\Unknown*redacted book name*(8)*redacted book name*- Unknown.pdf with error:
> > DeDRM v10.0.2: Trying to decrypt _redacted book name_- Unknown.pdf
> > DeDRM v10.0.2: _redacted book name_- Unknown.pdf is a PDF ebook with encryption EBX_HANDLER
> > DeDRM v10.0.2: _redacted book name_- Unknown.pdf is a PDF ebook (EBX) for UUID 1d72bb29-b23b-4659-b640-439a56124aec
> > Syntax Error: Couldn't find the 'EBX_HANDLER' security handler
> > pdfinfo errored out with return code: 1
> > InputFormatPlugin: PDF Input running
> > on C:\Users*redacted my name*\Calibre Library\Unknown*redacted book name*(8)*redacted book name*- Unknown.pdf
> > Failed to run pipe worker with command: from calibre.srv.render_book import viewer_main; viewer_main()
> > Traceback (most recent call last):
> > File "calibre\customize\ui.py", line 432, in get_file_type_metadata
> > File "calibre\customize\builtins.py", line 323, in get_metadata
> > File "calibre\ebooks\metadata\pdf.py", line 124, in get_metadata
> > ValueError: Could not read info dict from PDF
> > Running file type plugin DeDRM failed with traceback:
> > Traceback (most recent call last):
> > File "encodings\hex_codec.py", line 19, in hex_decode
> > ValueError: string argument should contain only ASCII characters
> > The above exception was the direct cause of the following exception:
> > Traceback (most recent call last):
> > File "calibre\customize\ui.py", line 175, in _run_filetype_plugins
> > File "calibre_plugins.dedrm.**init**", line 1050, in run
> > File "calibre_plugins.dedrm.**init**", line 917, in PDFDecrypt
> > File "calibre_plugins.dedrm.**init**", line 689, in PDFIneptDecrypt
> > ValueError: decoding with 'hex' codec failed (ValueError: string argument should contain only ASCII characters)
> > Traceback (most recent call last):
> > File "runpy.py", line 194, in _run_module_as_main
> > File "runpy.py", line 87, in _run_code
> > File "site.py", line 82, in 
> > File "site.py", line 77, in main
> > File "site.py", line 49, in run_entry_point
> > File "calibre\utils\ipc\worker.py", line 196, in main
> > File "", line 1, in 
> > File "calibre\srv\render_book.py", line 823, in viewer_main
> > File "calibre\srv\render_book.py", line 814, in render_for_viewer
> > File "calibre\srv\render_book.py", line 791, in render
> > File "calibre\ebooks\oeb\iterator\book.py", line 62, in extract_book
> > File "calibre\customize\conversion.py", line 242, in **call**
> > File "calibre\ebooks\conversion\plugins\pdf_input.py", line 51, in convert
> > File "calibre\ebooks\pdf\pdftohtml.py", line 82, in pdftohtml
> > calibre.ebooks.ConversionError: pdftohtml failed with return code: 1
> > Syntax Error: Couldn't find the 'EBX_HANDLER' security handler

Okay, I've solved it. You need to go into Calibre preferences, and go into the plugin's preferences and delete your Adobe Digital Editions key. It seems that if you add it yourself, you get this ascii error. If you try adding a DRM book but you don't have any keys added, the plugin will automatically do some automated search for the correct key (that will not give you an ascii error) and the DRM will be gone.

In other words, it's the "Create new Adobe Digital Editions Key" that is causing this ascii error; it's fetching some errored key. You just need to let the plugin sort itself out automatically.

ElleKayEm commented:
@TheWranger Ok, was surprised you were having trouble with an Internet Archive book since as far as I'm aware they don't use (or at least don't exclusively use) the newer Adobe DRM.

TheWranger commented:
> @TheWranger Ok, was surprised you were having trouble with an Internet Archive book since as far as I'm aware they don't use (or at least don't exclusively use) the newer Adobe DRM.

It probably isn't, I just didn't get around to trying it with the old version of DE

ElleKayEm commented:
> > @TheWranger Ok, was surprised you were having trouble with an Internet Archive book since as far as I'm aware they don't use (or at least don't exclusively use) the newer Adobe DRM.
> 
> It probably isn't, I just didn't get around to trying it with the old version of DE

I've always just stuck with ADE 2 so I know it won't be.  So far it hasn't stopped working for any of the places I get ebooks.

x0thrmc-exe commented:
> Okay, I've solved it. You need to go into Calibre preferences, and go into the plugin's preferences and delete your Adobe Digital Editions key. It seems that if you add it yourself, you get this ascii error. If you try adding a DRM book but you don't have any keys added, the plugin will automatically do some automated search for the correct key (that will not give you an ascii error) and the DRM will be gone.

**OMG this worked!** I was skeptical based on how many times I deleted and reset everything, but I simply did your steps and the book opened in Calibre no problem. After months I finally have a no DRM copy of that book. Thanks!



> I've always just stuck with ADE 2 so I know it won't be. So far it hasn't stopped working for any of the places I get ebooks.

I also recently tried the same book i was having problems with in ADE 2 after finding an installer on archive.org. It would not open in ADE 2, giving some error about the key.

ElleKayEm commented:
> I also recently tried the same book i was having problems with in ADE 2 after finding an installer on archive.org. It would not open in ADE 2, giving some error about the key.

Probably is the newer Adobe DRM when that happens.
